### DISCUSSION TITLE
Swarm in 2021?

### DISCUSSION THREAD

**Discussion by labanmne** [BretFisher/ama#189] [Q&A]
Hey there Bret, big thanks for your nicely explained course!
So I just went through the first part of the swarm section, and did some reading after it. I totally get swarm and its really fun to use, but people say its outdated, ie. essentially dead now. Is it worth it to continue with the swarm section, or should I just jump into Kubernetes? 

Once again, thanks, and cheers!

**Comment by BretFisher** [comment]
Thanks for the question @dascar5. This is something I think about (and get asked) a lot.

First, Swarm isn't dead. You can see active development in the [SwarmKit repo PR's](https://github.com/docker/swarmkit/pulls?q=is%3Apr+sort%3Aupdated-desc) (albeit slow these last 3 years with only a few maintainers still active).

There are multiple things to consider when picking an orchestrator, like any tool.  One of those is "is it the most popular and most supported" and that's a hard no for Swarm. It's **not** what I see businesses choosing for new production container deployments, but that doesn't mean it has no use for anyone. It's still the easiest self-run container orchestrator when you have minimal requirements. Docker still rolls SwarmKit updates and new features into the Docker Engine and CLI (like [Jobs support in 2020](https://github.com/moby/moby/pull/40307), and eventually [CSI storage support](https://github.com/moby/moby/issues/39624).) So, here are some things to think about if you've got a reason to *not go Kubernetes*.

**First, do you need persistent storage and advanced networking that the Swarm defaults don't support?** Then you'll likely want to skip Swarm and go K8s. Portworkx (fee-based beyond 3-node dev cluster) is the only 3rd party storage that I know of still in development as a Docker volumes plugin. You'll usually want volumes that are off-server (or replicated) if you need to run databases in Swarm and need them to fail over to other nodes automatically. [RexRay](https://github.com/rexray/rexray) used to be great for this but is a dead project.

**This might change with the next big feature for SwarmKit: CSI support (Container Storage Interface) built by @dperny at Mirantis.** We don't have a date for when it'll be merged and released with Docker Engine, but you can follow along with the commits by searching [`[feature-volumes]` in the repo](https://github.com/docker/swarmkit/issues?q=sort%3Aupdated-desc+in%3Atitle+%5Bfeature-volumes%5D).

Swarm is still a great tool for small (under 10 nodes), simple clusters, or personal setups when you don't need volumes or many 3rd party tools and you still want to run your own servers rather than cloud-provided K8s (which makes K8s as easy to setup and maintain as your own Swarm).

After all that, if you have any of these below requirements, then you'll likely want to skip Swarm and go K8s for that specific scenario:

- You've already got K8s in the org (just stick with the same orchestrator)
- The company uses this cluster to make money and needs support options, and you need to feel comfortable putting your name behind your production solution
- You'll likely want 3rd party tools for management outside of Portainer that officially support your orchestrator
- You want the cloud to build and manage your cluster for you
- You think your complexity will outgrow Swarm's existing features
- You want off-server volume storage and don't want to pay for Portworx

BONUS: Things I recommend for any Swarm install:

- https://www.portainer.io - Web UI, node management, and so much more
- https://traefik.io/traefik/ - Traefik ingress proxy for managing Let's Encrypt, TLS, and more
- https://github.com/stefanprodan/swarmprom - for monitoring example (now archived but still useful)
- https://dockerswarm.rocks - For guides and tools
- https://github.com/swarm-pack - for GitOps and things
- https://portworx.com/use-case/docker-persistent-storage/ - Portworx as 3rd party storage driver
